File: cs/demo/Glacier2/callback/CallbackClient.cs
// **********************************************************************
//
// Copyright (c) 2003-2005 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

using Demo;
using System;
using System.Collections;

class CallbackClient : Ice.Application
{
    private static void menu()
    {
        Console.WriteLine(
            "usage:\n" +
            "t: send callback as twoway\n" +
            "o: send callback as oneway\n" +
            "O: send callback as batch oneway\n" +
            "f: flush all batch requests\n" +
	    "v: set/reset override context field\n" +
	    "F: set/reset fake category\n" +
            "s: shutdown server\n" +
            "x: exit\n" +
            "?: help\n");
    }

    public override int run(string[] args)
    {
	Ice.RouterPrx defaultRouter = communicator().getDefaultRouter();
	if(defaultRouter == null)
	{
	    Console.Error.WriteLine("no default router set");
	    return 1;
	}
	
	Glacier2.RouterPrx router = Glacier2.RouterPrxHelper.checkedCast(defaultRouter);
	if(router == null)
	{
	    Console.Error.WriteLine("configured router is not a Glacier2 router");
	    return 1;
	}

	while(true)
	{
	    Console.WriteLine("This demo accepts any user-id / password combination.");

	    try
	    {
		String id;
		Console.Write("user id: ");
		Console.Out.Flush();
		id = Console.In.ReadLine();
		
		String pw;
		Console.Write("password: ");
		Console.Out.Flush();
		pw = Console.In.ReadLine();
		
		try
		{
		    router.createSession(id, pw);
		    break;
		}
		catch(Glacier2.PermissionDeniedException ex)
		{
		    Console.Write("permission denied:\n" + ex.reason);
		}
		catch(Glacier2.CannotCreateSessionException ex)
		{
		    Console.Write("cannot create session:\n" + ex.reason);
		}
	    }
            catch(System.Exception ex)
            {
                Console.Error.WriteLine(ex);
            }
	}

	String category = router.getServerProxy().ice_getIdentity().category;
	Ice.Identity callbackReceiverIdent = new Ice.Identity();
	callbackReceiverIdent.name = "callbackReceiver";
	callbackReceiverIdent.category = category;
	Ice.Identity callbackReceiverFakeIdent = new Ice.Identity();
	callbackReceiverFakeIdent.name = "callbackReceiver";
	callbackReceiverFakeIdent.category = "fake";

        Ice.Properties properties = communicator().getProperties();
        const string proxyProperty = "Callback.Proxy";
        String proxy = properties.getProperty(proxyProperty);
        if(proxy.Length == 0)
        {
            Console.Error.WriteLine("property `" + proxyProperty + "' not set");
            return 1;
        }

        Ice.ObjectPrx @base = communicator().stringToProxy(proxy);
        CallbackPrx twoway = CallbackPrxHelper.checkedCast(@base);
        CallbackPrx oneway = CallbackPrxHelper.uncheckedCast(twoway.ice_oneway());
        CallbackPrx batchOneway = CallbackPrxHelper.uncheckedCast(twoway.ice_batchOneway());

        Ice.ObjectAdapter adapter = communicator().createObjectAdapter("Callback.Client");
        adapter.add(new CallbackReceiverI(), callbackReceiverIdent);
        adapter.add(new CallbackReceiverI(), callbackReceiverFakeIdent);
        adapter.activate();

        CallbackReceiverPrx twowayR = CallbackReceiverPrxHelper.uncheckedCast(
	    adapter.createProxy(callbackReceiverIdent));
        CallbackReceiverPrx onewayR = CallbackReceiverPrxHelper.uncheckedCast(twowayR.ice_oneway());

        menu();

        string line = null;
	string @override = null;
	bool fake = false;
        do
        {
            try
            {
                Console.Write("==> ");
                Console.Out.Flush();
                line = Console.In.ReadLine();
                if(line == null)
                {
                    break;
                }
                if(line.Equals("t"))
                {
		    Ice.Context context = new Ice.Context();
		    context["_fwd"] = "t";
		    if(@override != null)
		    {
			context["_ovrd"] = @override;
		    }
                    twoway.initiateCallback(twowayR, context);
                }
                else if(line.Equals("o"))
                {
		    Ice.Context context = new Ice.Context();
		    context["_fwd"] = "o";
		    if(@override != null)
		    {
			context["_ovrd"] = @override;
		    }
                    oneway.initiateCallback(onewayR, context);
                }
                else if(line.Equals("O"))
                {
		    Ice.Context context = new Ice.Context();
		    context["_fwd"] = "O";
		    if(@override != null)
		    {
			context["_ovrd"] = @override;
		    } 
		    batchOneway.initiateCallback(onewayR, context);
                }
                else if(line.Equals("f"))
                {
		    communicator().flushBatchRequests();
                }
		else if(line.Equals("v"))
		{
		    if(@override == null)
		    {
			@override = "some_value";
			Console.WriteLine("override context field is now `" + @override + "'");
		    }
		    else
		    {
			@override = null;
			Console.WriteLine("override context field is empty");
		    }
		}
		else if(line.Equals("F"))
		{
		    fake = !fake;
		    
		    if(fake)
		    {
			twowayR = CallbackReceiverPrxHelper.uncheckedCast(
			    twowayR.ice_newIdentity(callbackReceiverFakeIdent));
			onewayR = CallbackReceiverPrxHelper.uncheckedCast(
			    onewayR.ice_newIdentity(callbackReceiverFakeIdent));
		    }
		    else
		    {
			twowayR = CallbackReceiverPrxHelper.uncheckedCast(
			    twowayR.ice_newIdentity(callbackReceiverIdent));
			onewayR = CallbackReceiverPrxHelper.uncheckedCast(
			    onewayR.ice_newIdentity(callbackReceiverIdent));
		    }
		    
		    Console.WriteLine("callback receiver identity: " + 
				      Ice.Util.identityToString(twowayR.ice_getIdentity()));
		}
                else if(line.Equals("s"))
                {
                    twoway.shutdown();
                }
                else if(line.Equals("x"))
                {
                    // Nothing to do
                }
                else if(line.Equals("?"))
                {
                    menu();
                }
                else
                {
                    Console.WriteLine("unknown command `" + line + "'");
                    menu();
                }
            }
            catch(System.Exception ex)
            {
                Console.Error.WriteLine(ex);
            }
        }
        while(!line.Equals("x"));

        return 0;
    }
}

File: vb/demo/Glacier2/callback/CallbackClient.vb
'  **********************************************************************
' 
'  Copyright (c) 2003-2005 ZeroC, Inc. All rights reserved.
' 
'  This copy of Ice is licensed to you under the terms described in the
'  ICE_LICENSE file included in this distribution.
' 
'  **********************************************************************

Imports glacier2demo.Demo
Imports System
Imports System.Collections

Class CallbackClient
    Inherits Ice.Application

    Private Sub menu()
        Console.Out.WriteLine("usage:")
        Console.Out.WriteLine("t: send callback as twoway")
        Console.Out.WriteLine("o: send callback as oneway")
        Console.Out.WriteLine("O: send callback as batch oneway")
        Console.Out.WriteLine("f: flush all batch requests")
        Console.Out.WriteLine("v: set/reset override context field")
        Console.Out.WriteLine("F: set/reset fake category")
        Console.Out.WriteLine("s: shutdown server")
        Console.Out.WriteLine("x: exit")
        Console.Out.WriteLine("?: help")
    End Sub

    Public Overloads Overrides Function run(ByVal args() As String) As Integer
        Dim defaultRouter As Ice.RouterPrx = communicator().getDefaultRouter()
        If defaultRouter Is Nothing Then
            Console.Error.WriteLine("no default router set")
            Return 1
        End If

        Dim router As Glacier2.RouterPrx = Glacier2.RouterPrxHelper.checkedCast(defaultRouter)
        If router Is Nothing Then
            Console.Error.WriteLine("configured router is not a Glacier2 router")
            Return 1
        End If

        While True
            Console.WriteLine("This demo accepts any user-id / password combination.")

            Try
                Dim id As String
                Console.Write("user id: ")
                Console.Out.Flush()
                id = Console.In.ReadLine()

                Dim pw As String
                Console.Write("password: ")
                Console.Out.Flush()
                pw = Console.In.ReadLine()

                Try
                    router.createSession(id, pw)
                    Exit While
                Catch ex As Glacier2.PermissionDeniedException
                    Console.Write("permission denied:\n" & ex.reason)
                Catch ex As Glacier2.CannotCreateSessionException
                    Console.Write("cannot create session:\n" & ex.reason)
                End Try
            Catch ex As System.Exception
                Console.Error.WriteLine(ex)
            End Try
        End While

        Dim category As String = router.getServerProxy().ice_getIdentity().category
        Dim callbackReceiverIdent As Ice.Identity = New Ice.Identity
        callbackReceiverIdent.name = "callbackReceiver"
        callbackReceiverIdent.category = category
        Dim callbackReceiverFakeIdent As Ice.Identity = New Ice.Identity
        callbackReceiverFakeIdent.name = "callbackReceiver"
        callbackReceiverFakeIdent.category = "fake"

        Dim properties As Ice.Properties = communicator().getProperties()
        Const proxyProperty As String = "Callback.Proxy"
        Dim proxy As String = properties.getProperty(proxyProperty)
        If proxy.Length = 0 Then
            Console.Error.WriteLine("property `" & proxyProperty & "' not set")
            Return 1
        End If

        Dim base As Ice.ObjectPrx = communicator().stringToProxy(proxy)
        Dim twoway As CallbackPrx = CallbackPrxHelper.checkedCast(base)
        Dim oneway As CallbackPrx = CallbackPrxHelper.uncheckedCast(twoway.ice_oneway())
        Dim batchOneway As CallbackPrx = CallbackPrxHelper.uncheckedCast(twoway.ice_batchOneway())

        Dim adapter As Ice.ObjectAdapter = communicator().createObjectAdapter("Callback.Client")
        adapter.add(New CallbackReceiverI, callbackReceiverIdent)
        adapter.add(New CallbackReceiverI, callbackReceiverFakeIdent)
        adapter.activate()

        Dim twowayR As CallbackReceiverPrx = CallbackReceiverPrxHelper.uncheckedCast(adapter.createProxy(callbackReceiverIdent))
        Dim onewayR As CallbackReceiverPrx = CallbackReceiverPrxHelper.uncheckedCast(twowayR.ice_oneway())

        menu()

        Dim line As String = Nothing
        Dim override As String = Nothing
        Dim fake As Boolean = False
        Do
            Try
                Console.Write("==> ")
                Console.Out.Flush()
                line = Console.In.ReadLine()
                If line Is Nothing Then
                    Exit Try
                End If
                If line.Equals("t") Then
                    Dim context As Ice.Context = New Ice.Context
                    context("_fwd") = "t"
                    If Not override Is Nothing Then
                        context("_ovrd") = override
                    End If
                    twoway.initiateCallback(twowayR, context)
                ElseIf line.Equals("o") Then
                    Dim context As Ice.Context = New Ice.Context
                    context("_fwd") = "o"
                    If override Is Nothing Then
                        context("_ovrd") = override
                    End If
                    oneway.initiateCallback(onewayR, context)
                ElseIf line.Equals("O") Then
                    Dim context As Ice.Context = New Ice.Context
                    context("_fwd") = "O"
                    If Not override Is Nothing Then
                        context("_ovrd") = override
                    End If
                    batchOneway.initiateCallback(onewayR, context)
                ElseIf line.Equals("f") Then
                    communicator().flushBatchRequests()
                ElseIf line.Equals("v") Then
                    If override Is Nothing Then
                        override = "some_value"
                        Console.WriteLine("override context field is now `" & override & "'")
                    Else
                        override = Nothing
                        Console.WriteLine("override context field is empty")
                    End If
                ElseIf line.Equals("F") Then
                    fake = Not fake

                    If fake Then
                        twowayR = CallbackReceiverPrxHelper.uncheckedCast(twowayR.ice_newIdentity(callbackReceiverFakeIdent))
                        onewayR = CallbackReceiverPrxHelper.uncheckedCast(onewayR.ice_newIdentity(callbackReceiverFakeIdent))
                    Else
                        twowayR = CallbackReceiverPrxHelper.uncheckedCast(twowayR.ice_newIdentity(callbackReceiverIdent))
                        onewayR = CallbackReceiverPrxHelper.uncheckedCast(onewayR.ice_newIdentity(callbackReceiverIdent))
                    End If

                    Console.WriteLine("callback receiver identity: " & Ice.Util.identityToString(twowayR.ice_getIdentity()))
                ElseIf line.Equals("s") Then
                    twoway.shutdown()
                ElseIf line.Equals("x") Then
                    '  Nothing to do
                ElseIf line.Equals("?") Then
                    menu()
                Else
                    Console.WriteLine("unknown command `" & line & "'")
                    menu()
                End If
            Catch ex As System.Exception
                Console.Error.WriteLine(ex)
            End Try
        Loop While Not line.Equals("x")

        Return 0
    End Function

End Class
